src/dictionary.c

```diff
@@ -160,6 +158,37 @@ void dict_put(struct dictionary* d, uint32_t hash, R_len_t i) {
   d->used++;
 }
 
+// Assume worst case, that every value is distinct, aiming for a load factor
+// of at most 77%. We round up to power of 2 to ensure quadratic probing
+// strategy works. Maximum power of 2 we can store in a uint32_t is 2^31,
+// as 2^32 is 1 greater than the max uint32_t value, so we clamp sizes that
+// would result in 2^32 to INT_MAX to ensure that our maximum ceiling value
+// is only 2^31. This will increase the load factor above 77% for `x` with
+// length greater than 1653562409 (2147483648 * .77), but it ensures that
+// it can run.
+static inline
+uint32_t dict_key_size(SEXP x) {
+  const R_len_t x_size = vec_size(x);
+
+  if (x_size > R_LEN_T_MAX) {
```

Comment:
Needs a comment that this is in prevision of a switch to long vector sizes.

Reply:
04e28ad8f52972f479f7dad31f3061337e13c8f9